File: test/trigonometric.js
// ------------------------------------------------------------------------
// Copyright (c) 2019-2024 Alexandre Bento Freire. All rights reserved.
// Licensed under the MIT License
// ------------------------------------------------------------------------

/**
 * Mocha tests for numnjstoix trigonometric functions.
 */

// ------------------------------------------------------------------------
//                               Requires
// ------------------------------------------------------------------------

import { numnjstoix } from '../main.js';
import * as shared from './shared.js';

// ------------------------------------------------------------------------
//                               Test Cases
// ------------------------------------------------------------------------

describe('Trigonometric', function () {

  shared.addTransformCase(describe, 'sin', Math.sin, numnjstoix.sin, true);
  shared.addTransformCase(describe, 'cos', Math.cos, numnjstoix.cos, true);
  shared.addTransformCase(describe, 'tan', Math.tan, numnjstoix.tan, true);

  shared.addTransformCase(describe, 'arcsin', Math.asin, numnjstoix.arcsin, true);
  shared.addTransformCase(describe, 'arccos', Math.acos, numnjstoix.arccos, true);
  shared.addTransformCase(describe, 'arctan', Math.atan, numnjstoix.arctan, true);

  shared.addTransformCase(describe, 'degrees', (v) => v * (180 / Math.PI), numnjstoix.degrees, true);
  shared.addTransformCase(describe, 'radians', (v) => v / (180 / Math.PI), numnjstoix.radians, true);
});

File: src/trigonometric.cpp
// ------------------------------------------------------------------------
// Copyright (c) 2019-2024 Alexandre Bento Freire. All rights reserved.
// Licensed under the MIT License
// ------------------------------------------------------------------------

// ------------------------------------------------------------------------
//                               Includes
// ------------------------------------------------------------------------

#define _USE_MATH_DEFINES
#include "numnjs.h"
#include <math.h>
#include <node.h>

namespace numnjs {

// ------------------------------------------------------------------------
//                               Actual Functions
// ------------------------------------------------------------------------

/* clang-format off */
transfCalc1DblFunc(sinFunc, sin)
transfCalc1DblFunc(cosFunc, cos)
transfCalc1DblFunc(tanFunc, tan)

transfCalc1DblFunc(arcsinFunc, asin)
transfCalc1DblFunc(arccosFunc, acos)
transfCalc1DblFunc(arctanFunc, atan)

transfCalc1DblExpr(degreesFunc,  v * (180 * M_1_PI))
transfCalc1DblExpr(radiansFunc,  v / (180 * M_1_PI))
/* clang-format on */

// ------------------------------------------------------------------------
//                               Initialize
// ------------------------------------------------------------------------

void initializeTrigonometric(v8exports exports) {

  NODE_SET_METHOD(exports, "sin", sinFunc);
  NODE_SET_METHOD(exports, "cos", cosFunc);
  NODE_SET_METHOD(exports, "tan", tanFunc);

  NODE_SET_METHOD(exports, "arcsin", arcsinFunc);
  NODE_SET_METHOD(exports, "arccos", arccosFunc);
  NODE_SET_METHOD(exports, "arctan", arctanFunc);

  NODE_SET_METHOD(exports, "degrees", degreesFunc);
  NODE_SET_METHOD(exports, "radians", radiansFunc);
}

} // namespace numnjs